docs/devel/e2e-tests.md

```diff
@@ -232,6 +232,102 @@ stale permissions can cause problems.
 
   - `sudo iptables -F`, clear ip tables rules left by the kube-proxy.
 
+### Federation e2e tests
+
+By default, `e2e.go` provisions a single Kubernetes cluster, and any `Feature:Federation` ginkgo tests will be skipped.
+
+Federation e2e testing involve bringing up multiple "underlying" Kubernetes clusters,
+and deploying the federation control plane as a Kubernetes application on the underlying clusters.
+
+The federation e2e tests are still managed via `e2e.go`, but require some extra configuration items.
+
+#### Configuring federation e2e tests
+
+The following environment variables will enable federation e2e building, provisioning and testing.
+
+```sh
+$ export FEDERATION=true
+$ export E2E_ZONES="us-central1-a us-central1-b us-central1-f"
+```
+
+A Kubernetes cluster will be provisioned in each zone listed in `E2E_ZONES`. A zone can only appear once in the `E2E_ZONES` list.
+
+#### Image Push Repository
+
+Next, specify the docker repository where your ci images will be pushed.
+
+If you are using `KUBERNETES_PROVIDER=gce`, you can simply set your push repo base based on your project name, and the
+necessary repositories will be auto-created when you first push your container images:
+
+```sh
+$ export FEDERATION_PUSH_REPO_BASE="gcr.io/${GCE_PROJECT_NAME}"
+```
+
+Otherwise, if the provider is not gce, you'll be responsible for creating and managing access to the repositories manually.
+
+```sh
+$ export FEDERATION_PUSH_REPO_BASE="quay.io/colin_hom"
+```
+
+Given this example, the `federation-apiserver` container image will be pushed to the repository `quay.io/colin_hom/federation-apiserver`.
+
+The docker client on the machine running `e2e.go` must have push access for the following pre-existing repositories:
+
+* `${FEDERATION_PUSH_REPO_BASE}/federation-apiserver`
+* `${FEDERATION_PUSH_REPO_BASE}/federation-controller-manager`
+
+These repositories must allow public read access, as the e2e node docker daemons will not have any credentials.
+
+#### Build
+* Compile the binaries and build container images:
+
+  ```sh
+  $ KUBE_RELEASE_RUN_TESTS=n KUBE_FASTBUILD=true go run hack/e2e.go -v -build
+  ```
+
+* Push the federation container images
+
+  ```sh
+  $ build/push-federation-images.sh
+  ```
+
+#### Deploy federation control plane
+
+The following command will create the underlying Kubernetes clusters in each of `E2E_ZONES`, and then provision the
+federation control plane on one of those clusters:
+
+```sh
+$ go run hack/e2e.go -v -up
+```
+
+**Note for e2e infra developers**: If you're hacking around with the federation control plane deployment itself,
+you can quickly re-deploy the federation control plane Kubernetes manifests without tearing any resources down.
+
+To re-deploy the federation control plane after running `-up` for the first time:
+
+```sh
+$ federation/cluster/federated-up.sh
```

Comment:
Now that I think about it, instead of generating these individual kubeconfig files at cluster turn up time (in e2e.go --up) and then using them in federation-up, we could have extracted the info for each k8s cluster from the combined kubeconfig in federation-up. That way federation-up.sh will work even if e2e.go --up was run without FEDERATION=true.
Its also not clear how useful running federation-up after a normal e2e.go -up is going to be. People will want multiple k8s clusters. So I am going to let the current code be as is. We can change it later.